core/src/apps/cardano/sign_tx.py

```diff
@@ -300,32 +293,33 @@ def _validate_metadata(metadata: Optional[bytes]) -> None:
 
 def _get_signed_tx_chunks(
     cborized_tx: CborizedSignedTx,
-) -> Iterator[Tuple[bytes, bool]]:
-    remaining_len = 0
-    for token in cbor.encode_chunked(cborized_tx):
-        remaining_len += len(token)
-
-    chunks = cbor.encode_chunked(cborized_tx)
+) -> Iterator[bytearray]:
+    encoded_chunks = cbor.encode_chunked(cborized_tx)
     current_chunk_view = utils.BufferReader(b"")
-    while remaining_len > 0:
-        payload_len = min(remaining_len, MAX_TX_CHUNK_SIZE)
-        payload_writer = utils.BufferWriter(bytearray(payload_len))
 
-        while payload_writer.offset < payload_len:
+    signed_tx_chunk = writers.empty_bytearray(MAX_TX_CHUNK_SIZE)
+
+    has_encoded_chunks = True
+    while has_encoded_chunks:
+        signed_tx_chunk[:] = bytes()
+
+        while len(signed_tx_chunk) < MAX_TX_CHUNK_SIZE:
             if current_chunk_view.remaining_count() == 0:
-                current_chunk_view = utils.BufferReader(next(chunks))
+                try:
+                    encoded_chunk = next(encoded_chunks)
+                    current_chunk_view = utils.BufferReader(encoded_chunk)
+                except StopIteration:
+                    has_encoded_chunks = False
```

Comment:
It seems quite awkward to me to "bail out" from the loop like this. also if there are no more encoded chunks to yield at this point, isn't it enough to flush the current buffer?

I tried rewriting the logic like this and seems to work fine:

```python
def _get_signed_tx_chunks(
    cborized_tx: CborizedSignedTx,
) -> Iterator[bytearray]:
    chunk_buffer = writers.empty_bytearray(MAX_SERIALIZED_TX_CHUNK_SIZE)

    chunks = cbor.encode_chunked(cborized_tx)
    current_chunk_view = utils.BufferReader(b"")
    try:
        while True:
            if current_chunk_view.remaining_count() == 0:
                current_chunk_view = utils.BufferReader(next(chunks))

            num_bytes_to_write = min(
                current_chunk_view.remaining_count(),
                MAX_SERIALIZED_TX_CHUNK_SIZE - len(chunk_buffer),
            )
            chunk_buffer.extend(current_chunk_view.read(num_bytes_to_write))

            if len(chunk_buffer) >= MAX_SERIALIZED_TX_CHUNK_SIZE:
                yield chunk_buffer
                chunk_buffer[:] = bytes()
    except StopIteration:
        yield chunk_buffer
```

Reply:
I agree it was a bit awkward, but I wanted to avoid having the yield in two places and I couldn't figure a nice way to do that. What you propose does use two yields, but it's pretty clear what's happening. Nice :+1: